File: 1679-shortest-subarray-to-be-removed-to-make-array-sorted/shortest-subarray-to-be-removed-to-make-array-sorted.py
class Solution:
    def findLengthOfShortestSubarray(self, arr: List[int]) -> int:
        N = len(arr)
        # remove prefix
        l, r = 0, N - 1
        while r > l and arr[r - 1] <= arr[r]:
            r -= 1

        res = r
        while l < r:
            # move r to to make the window valid
            while r < N and arr[r] < arr[l]:
                r += 1
            res = min(res, r - l - 1)
            if arr[l] > arr[l + 1]:     # left break condition
                break
            l += 1

        return res

File: 1679-shortest-subarray-to-be-removed-to-make-array-sorted/shortest-subarray-to-be-removed-to-make-array-sorted.java
/*
    1   2   3   10  4   2   3   5
 */

class Solution {
    public int findLengthOfShortestSubarray(int[] arr) {
        // for every left pointer, find the right pointer
        // as the left pointer increases, right should increase or remain the same
        // if right does not increase then the deletion window becomes shorter

        int l = 0;
        int r = arr.length - 1;
        while (l < r && arr[r - 1] <= arr[r]) {
            r--;
        }

        int shortest = r;
        while (l < r) {
            while (r < arr.length && arr[l] > arr[r]) {
                r++;
            }
            shortest = Math.min(shortest, r - l - 1);
            // System.out.println(l + " " + r + " " + shortest);
            if (arr[l] > arr[l + 1]) {
                break;
            }
            l++;
        }

        return shortest;
    }
}